xarray/namedarray/core.py

```diff
@@ -39,90 +66,279 @@
+    and attributes which describe a single Array.
+    Numeric operations on this object implement array broadcasting and
+    dimension alignment based on dimension names,
+    rather than axis order.
+
+
+    Parameters
+    ----------
+    dims : str or iterable of str
+        Name(s) of the dimension(s).
+    data : T_DuckArray
+        The actual data that populates the array. Should match the
+        shape specified by `dims`.
+    attrs : dict, optional
+        A dictionary containing any additional information or
+        attributes you want to store with the array.
+        Default is None, meaning no attributes will be stored.
+
+    Raises
+    ------
+    ValueError
+        If the `dims` length does not match the number of data dimensions (ndim).
+
+
+    Examples
+    --------
+    >>> data = np.array([1.5, 2, 3], dtype=float)
+    >>> narr = NamedArray(("x",), data, {"units": "m"})  # TODO: Better name than narr?
+    """
 
     __slots__ = ("_data", "_dims", "_attrs")
 
-    _data: T_DuckArray
-    _dims: Dims
+    _data: duckarray[Any, _DType_co]
+    _dims: _Dims
     _attrs: dict[Any, Any] | None
 
     def __init__(
         self,
-        dims: DimsInput,
-        data: T_DuckArray | np.typing.ArrayLike,
-        attrs: AttrsInput = None,
-        fastpath: bool = False,
+        dims: _DimsLike,
+        data: duckarray[Any, _DType_co],
+        attrs: _AttrsLike = None,
     ):
+        self._data = data
+        self._dims = self._parse_dimensions(dims)
+        self._attrs = dict(attrs) if attrs else None
+
+    def __init_subclass__(cls, **kwargs: Any) -> None:
+        if NamedArray in cls.__bases__ and (cls._new == NamedArray._new):
+            # Type hinting does not work for subclasses unless _new is overriden with
+            # the correct class.
+            raise TypeError(
+                "Subclasses of `NamedArray` must override the `_new` method."
+            )
+        super().__init_subclass__(**kwargs)
+
+    @overload
+    def _new(
+        self,
+        dims: _DimsLike | Default = ...,
+        data: duckarray[_ShapeType, _DType] = ...,
+        attrs: _AttrsLike | Default = ...,
+    ) -> NamedArray[_ShapeType, _DType]:
+        ...
+
+    @overload
+    def _new(
+        self,
+        dims: _DimsLike | Default = ...,
+        data: Default = ...,
+        attrs: _AttrsLike | Default = ...,
+    ) -> NamedArray[_ShapeType_co, _DType_co]:
+        ...
+
+    def _new(
+        self,
+        dims: _DimsLike | Default = _default,
+        data: duckarray[Any, _DType] | Default = _default,
+        attrs: _AttrsLike | Default = _default,
+    ) -> NamedArray[_ShapeType, _DType] | NamedArray[_ShapeType_co, _DType_co]:
+        return _new(self, dims, data, attrs)
+
+    def _replace(
+        self,
+        dims: _DimsLike | Default = _default,
+        data: duckarray[_ShapeType_co, _DType_co] | Default = _default,
+        attrs: _AttrsLike | Default = _default,
+    ) -> Self:
```

Comment:
That's not really correct since the dtype and shape might change.

Not sure on how to type this correctly or if it is even possible to type this with the current type system in python.

Reply:
My idea about this is:
* Use `self._replace` if shape and dtype is the same.
* Use `self._new` if shape or dtype doesn't match.

Then I think it'll behave as expected. 

It kind of has to look like this because we subclass NamedArray in Variable. We would need Higher-Kinded TypeVars if we want to get this really correct.

Reply:
Yeah, that's a rough limitation. I assume numpy has similar problems? How are they typing such functions like `astype`?

Reply:
Not in a subclass-friendly way:
https://github.com/numpy/numpy/blob/14bb214bca49b167abc375fa873466a811e62102/numpy/__init__.pyi#L1776-L1793